Sources/SQLKit/Builders/SQLUnionBuilder.swift

```diff
@@ -0,0 +1,25 @@
+public final class SQLUnionBuilder: SQLQueryBuilder {
+    public var query: SQLExpression {
+        return self.union
+    }
+
+    public var union: SQLUnion
+    public var database: SQLDatabase
+
+    public init(on database: SQLDatabase,
+                _ args: [SQLSelectBuilder],
+                all: Bool = false) {
+        self.union = .init(args.map(\.select), all: all)
+        self.database = database
+    }
+}
+
+extension SQLDatabase {
+    public func union(_ args: SQLSelectBuilder...) -> SQLUnionBuilder {
```

Comment:
Elsewhere that we deal with subqueries, we tend to use this kind of construct instead of accepting builders as parameters:
```swift
extension SQLSelectBuilder {
    public func union(select builder: (SQLSelectBuilder) -> SQLSelectBuilder) -> SQLUnionBuilder {
        return .init(on: self, [self.select, predicate(.init(on: self.database))])
    }
}
```
With usage then looking like this:
```swift
database.select()
    .columns(whatever)
    .from(whatever)
.union(select: { $0
    .columns(whatever)
    .from(whatever)
})
// And an extension on SQLUnionBuilder would allow chaining this way instead of nesting closures:
.unionAll(select: { $0
    .columns(whatever)
    .from(whatever)
})
```
It could also be set up to be used like this:
```swift
database.union(select: {
    $0.columns().etcetera()
}).union(select: {
    $0.etcetera()
}).unionAll(select: {
    $0.etcetera()
})
```
Or both, which is probably what I'd go with if I were implementing this syntax.

Whether this is better or worse than the way it is here of just taking a variadic array of select builders is of course a matter of opinion; I'm open to arguments in both directions.

Reply:
Interesting, it hadn't even occurred to me that this would be an option. I think I'd still lean towards `union(_ args: SQLSelectBuilder...`, because the operands are "equal".

What I mean is that while `.union(select: ...)` more closely matches the order of the SQL syntax, it does introduce some nesting (for lack of a better term) that the SQL syntax doesn't have. Or said differently, to me `a union b` maps more naturally to `union(a, b)` than `a.union(b)`.

This is just a personal preference though, happy to adopt whatever is preferred!

Reply:
Yeah, at the very least I'd definitely lean towards the second example, where `union(select:)` etc. are methods on both `SQLDatabase` and `SQLUnionBuilder`, but not `SQLSelectBuilder`, and I also agree that your original proposal is more straightforward - my main argument for the closure-based syntax is that it matches other usages in `SQLKit`, and I tend to want to avoid making this layer any more finicky to use than it already is 😅. Of course, at the same time, "straightforward" may be better than "consistent", so all in all I'm not sure what's best either.

Reply:
So if I understand correctly, this would transform the test from

```swift
        try db.union(
            db.select()
                .column("id")
                .from("t1")
                .where("f1", .equal, "foo")
                .limit(1),
            db.select()
                .column("id")
                .from("t2")
                .where("f2", .equal, "bar")
                .limit(2),
            db.select()
                .column("id")
                .from("t3")
                .where("f3", .equal, "baz")
                .limit(3)
        ).run().wait()
```

to

```swift
        try db
	.union(select: {
            $0.column("id")
                .from("t1")
                .where("f1", .equal, "foo")
                .limit(1)
		})
	.union(select: {
            $0.column("id")
                .from("t2")
                .where("f2", .equal, "bar")
                .limit(2)
		})
	.union(select: {
            $0.column("id")
                .from("t3")
                .where("f3", .equal, "baz")
                .limit(3)
        }).run().wait()
```

We're using SQLKit directly (instead of Fluent) in a few places and I've not yet come across this kind of closure based API, I think. As an API user I'd definitely prefer the former (unsurprisingly, as it's the API I reached for coming from the other end!).

In particular, I'm testing the individual union queries (which can be quite complicated) and it's helpful that they're valid standalone queries (like the original SQL queries are just concatenations). For example:

- https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/blob/swift-5.5/Sources/App/Core/Search.swift#L248-L249
- https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/blob/swift-5.5/Sources/App/Core/Search.swift#L198
- https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/blob/swift-5.5/Tests/AppTests/SearchTests.swift#L70

I'm not sure that kind of composability could be preserved unless `union` operates on builders? I have to admit I haven't spent a lot of time investigating. I'm happy to adopt and deal with whatever the API will end up being, even if it means to add a little wrapper in SPI to make that feasible.

Reply:
I admit there's only one place where closures are used by `SQLKit` already - `SQLPredicateBuilder.where(group:)` and `orWhere(group:)`, e.g. `db.select()./* blah */.where(\Baz.baz != nil).where { $0.where(\Foo.foo == 1).orWhere(\Bar.bar == 2) }` -> `WHERE baz IS NOT NULL AND (Foo.foo=1 OR Bar.bar=2)`. I also used them in my proposed syntax for providing "`UPSERT`" support (the PR is stale but not abandoned: #90).

The closure-based syntax does ultimately still operate on `SQLSelectBuilder`s; the main benefit I see from the point of view of the API caller is that they're no longer manually invoking `db.select()` inside another `db.<something>()` call - one particular benefit of having the select builders created internally by the union builder is a user can no longer (accidentally or otherwise) do something like `db.union(someOtherDb.select().whatever())` (which doesn't make sense and wouldn't have whatever effect may have been intended - even done accidentally it'd just be confusion waiting to happen for code maintainers).

Reply:
So I have no clear winner in my mind. I agree that option a above looks nicer and seems like it would compose better but I'm definitely in favour of trying to enforce some type-safety around the DB being used etc. 

So in my mind - we see if it's possible to extract out the select builder into something that can be composed (outside of just pulling the closure out into a function - is this an option?) and if not pick option 1

Reply:
The select builder can be composed directly as long as it's the closure return value as well as its parameter so the closure can return a composite of its own. The call site, in turn, should be taking the `.select` of the return value rather than of the parameter, rather than assuming they're always the same as `SQLPredicateGroupBuilder` currently incorrectly does (this admittedly partially defeats the "prevent the user from mixing databases in the builders" benefit, but composite sub-builders are a very rare usage and one would have to go out of their way to make the error rather than it being relatively trivial).

Reply:
Let me show you really quickly a sketch of how I'd implement this...

```swift
public final class SQLUnionBuilder: SQLQueryBuilder {
     public var query: SQLExpression { self.union }

     public var union: SQLUnion
     public var database: SQLDatabase

     public init(on database: SQLDatabase, initialQuery: SQLSelect) {
         self.union = .init(initialQuery: initialQuery)
         self.database = database
     }
    
    public func union(distinct query: SQLSelect) -> SQLUnionBuilder {
        self.union.add(query, all: false)
        return self
    }
    
    public func union(all query: SQLSelect) -> SQLUnionBuilder {
        self.union.add(query, all: true)
        return self
    }
 }

public struct SQLUnion: SQLExpression {
    public var initialQuery: SQLSelect
    public var unions: [(SQLUnionJoiner, SQLSelect)]
    
    public init(initialQuery: SQLSelect, unions: [(SQLUnionJoiner, SQLSelect)] = []) {
        self.initialQuery = initialQuery
        self.unions = unions
    }
    
    public func add(_ query: SQLSelect, all: Bool) {
        self.unions.append((.init(all: all), query))
    }

    public func serialize(into serializer: inout SQLSerializer) {
        assert(!self.unions.isEmpty, "Serializing a union with only one query is invalid.")
        self.initialQuery.serialize(into: &serializer)
        self.unions.flatMap { [$0, $1] }.forEach( (item: SQLExpression) in item.serialize(into: &serializer) }
    }
}

extension SQLDatabase {
    public func union(_ predicate: (SQLSelectBuilder) -> SQLSelectBuilder) -> SQLUnionBuilder {
        return SQLUnionBuilder(on: self, initialQuery: predicate(.init(on: self)).select)
    }
}

extension SQLUnionBuilder {
    public func union(distinct predicate: (SQLSelectBuilder) -> SQLSelectBuilder) -> Self {
        return self.union(distinct: predicate(.init(on: self.database)).select)
    }

    public func union(all predicate: (SQLSelectBuilder) -> SQLSelectBuilder) -> Self {
        return self.union(all: predicate(.init(on: self.database)).select)
    }
}
```

Something along those lines, anyhow.

Reply:
This snippet:

```swift
    public func serialize(into serializer: inout SQLSerializer) {
        assert(!self.unions.isEmpty, "Serializing a union with only one query is invalid.")
        self.initialQuery.serialize(into: &serializer)
        self.unions.flatMap { [$0, $1] }.forEach( (item: SQLExpression) in item.serialize(into: &serializer) }
    }
```

didn't quite compile. I've come up with this replacement that passes the tests (will push a change in a minute):

```swift
    public func serialize(to serializer: inout SQLSerializer) {
        assert(!self.unions.isEmpty, "Serializing a union with only one query is invalid.")
        SQLGroupExpression(self.initialQuery).serialize(to: &serializer)
        self.unions
            .forEach { (joiner, select) in
                joiner.serialize(to: &serializer)
                SQLGroupExpression(select).serialize(to: &serializer)
            }
    }
```

Was that what you had in mind? It looks quite different, so I'm not sure.